src/components/views/settings/JoinRuleSettings.tsx

```diff
@@ -0,0 +1,268 @@
+}
+
+const JoinRuleSettings = ({ room, promptUpgrade, onError, beforeChange, closeSettingsFn }: IProps) => {
+    const cli = room.client;
+
+    const restrictedRoomCapabilities = SpaceStore.instance.restrictedJoinRuleSupport;
+    const roomSupportsRestricted = Array.isArray(restrictedRoomCapabilities?.support)
+        && restrictedRoomCapabilities.support.includes(room.getVersion());
+    const preferredRestrictionVersion = !roomSupportsRestricted && promptUpgrade
+        ? restrictedRoomCapabilities?.preferred
+        : undefined;
+
+    const disabled = !room.currentState.mayClientSendStateEvent(EventType.RoomJoinRules, cli);
+
+    const [content, setContent] = useLocalEcho<IJoinRuleEventContent>(
+        () => room.currentState.getStateEvents(EventType.RoomJoinRules, "")?.getContent(),
+        content => cli.sendStateEvent(room.roomId, EventType.RoomJoinRules, content, ""),
+        onError,
+    );
+
+    const { join_rule: joinRule } = content;
+    const restrictedAllowRoomIds = joinRule === JoinRule.Restricted
+        ? content.allow.filter(o => o.type === RestrictedAllowType.RoomMembership).map(o => o.room_id)
+        : undefined;
+
+    const editRestrictedRoomIds = async (): Promise<string[] | undefined> => {
+        let selected = restrictedAllowRoomIds;
+        if (!selected?.length && SpaceStore.instance.activeSpace) {
+            selected = [SpaceStore.instance.activeSpace.roomId];
+        }
+
+        const matrixClient = MatrixClientPeg.get();
+        const { finished } = Modal.createTrackedDialog('Edit restricted', '', ManageRestrictedJoinRuleDialog, {
+            matrixClient,
+            room,
+            selected,
+        }, "mx_ManageRestrictedJoinRuleDialog_wrapper");
+
+        const [roomIds] = await finished;
+        return roomIds;
+    };
+
+    const definitions: IDefinition<JoinRule>[] = [{
+        value: JoinRule.Invite,
+        label: _t("Private (invite only)"),
+        description: _t("Only invited people can join."),
+        checked: joinRule === JoinRule.Invite || (joinRule === JoinRule.Restricted && !restrictedAllowRoomIds?.length),
+    }, {
+        value: JoinRule.Public,
+        label: _t("Public"),
+        description: _t("Anyone can find and join."),
+    }];
+
+    if (roomSupportsRestricted || preferredRestrictionVersion || joinRule === JoinRule.Restricted) {
+        let upgradeRequiredPill;
+        if (preferredRestrictionVersion) {
+            upgradeRequiredPill = <span className="mx_SecurityRoomSettingsTab_upgradeRequired">
+                { _t("Upgrade required") }
+            </span>;
+        }
+
+        let description;
+        if (joinRule === JoinRule.Restricted && restrictedAllowRoomIds?.length) {
+            const shownSpaces = restrictedAllowRoomIds
+                .map(roomId => cli.getRoom(roomId))
+                .filter(room => room?.isSpaceRoom())
+                .slice(0, 4);
```

Comment:
suggestion (non-blocking): Could we add a bit  more context regarding that limit? Reading through this I'm unsure why we would want to limit the amount of shown spaces to the first four in that list?
It's also unclear to me what happens to the spaces that get discarded here